src/host-config/proxy.ts

```diff
@@ -0,0 +1,270 @@
+			rawRedsocksBlockMatch[1],
+			disallowedProxyFields,
+		);
+		const maybeProxyConfig = ProxyConfig.decode(rawRedsocksBlock);
+		if (isRight(maybeProxyConfig)) {
+			conf.redsocks = {
+				...maybeProxyConfig.right,
+			};
+		} else {
+			log.warn(Reporter.report(maybeProxyConfig).join('\n'));
+			// Allow the invalid values through for backwards compatibility
+			if (Object.keys(rawRedsocksBlock).length) {
+				conf.redsocks = rawRedsocksBlock as unknown as ProxyConfig;
+			}
+		}
+
+		return conf;
+	}
+
+	private static stringifyBlock(
+		label: string,
+		block: Record<string, any>,
+		readonlyConfigs: Record<string, any> = {},
+	): string {
+		const lines = Object.entries(readonlyConfigs)
+			.concat(
+				Object.entries(block).filter(
+					([key]) => !disallowedProxyFields.includes(key),
+				),
+			)
+			.map(([key, value]) => {
+				if (isAuthField(key)) {
+					// Add double quotes around login and password fields
+					value = `${value.startsWith('"') ? '' : '"'}${value}${value.endsWith('"') ? '' : '"'}`;
+				}
+				return `\t${key} = ${value};`;
+			});
+		return `${label} {\n${lines.join('\n')}\n}\n`;
+	}
+
+	/**
+	 * Given the raw contents of a block redsocks.conf file,
+	 * extract to a key-value object.
+	 */
+	private static parseBlock(
+		rawBlockConf: string,
+		unsupportedKeys: string[],
+	): Record<string, string> {
+		const parsedBlock: Record<string, string> = {};
+
+		// Split by newline and optional semicolon
+		for (const line of rawBlockConf.split(/;?\n/)) {
+			if (!line.trim().length) {
+				continue;
+			}
+			let [key, value] = line.split(/ *?= *?/).map((s) => s.trim());
+			// Don't parse unsupported keys
+			if (key && unsupportedKeys.some((k) => key.match(k))) {
+				continue;
+			}
+			if (key && value) {
+				if (isAuthField(key)) {
+					// Remove double quotes from login and password fields for readability
+					value = value.replace(/"/g, '');
+				}
+				parsedBlock[key] = value;
+			} else {
+				// Skip malformed lines
+				log.warn(
+					`Ignoring malformed redsocks.conf line ${isAuthField(key) ? `"${key}"` : `"${line.trim()}"`} due to missing key, value, or "="`,
+				);
+			}
+		}
+
+		return parsedBlock;
+	}
+}
+
+export async function readProxy(): Promise<HostProxyConfig | undefined> {
+	// Get and parse redsocks.conf
+	let rawConf: string | undefined;
+	try {
+		rawConf = await readFromBoot(redsocksConfPath, 'utf-8');
+	} catch (e: unknown) {
+		if (!isENOENT(e)) {
+			throw e;
+		}
+		return undefined;
+	}
+	const redsocksConf = RedsocksConf.parse(rawConf);
+
+	// Get and parse no_proxy
+	const noProxy = await readNoProxy();
+
+	// Build proxy object
+	const proxy = {
+		...redsocksConf.redsocks,
+		...(noProxy.length && { noProxy }),
+	};
+
+	// Assumes mandatory proxy config fields (type, ip, port) are present,
+	// even if they very well may not be. It is up to the user to ensure
+	// that all the necessary fields are present in the redsocks.conf file.
+	return proxy as HostProxyConfig;
```

Comment:
Maybe a way to do this is that the returned config here allows partials while the input config does not?

Reply:
The input config as in `setProxy`? 

`readProxy` just reads and returns whatever is from redsocks.conf, which could be anything as the user could easily preload an invalid file onto the device. I don't think there is much benefit to differentiating `Partial<HostProxyConfig>` vs `HostProxyConfig` in the return type, because we wouldn't do anything differently if the return was a partial -- if `type`, `ip`, or `port` aren't present, we can't do anything about that other than maybe log it. I really think it's the user's responsibility here to make sure any configuration in redsocks.conf is valid if preloaded.

Reply:
This relates to this [other comment](https://github.com/balena-os/balena-supervisor/pull/2329/files#r1657733759) about the contract of the function signature. Maybe we should be stricter and say, "if there is a field in the required list missing then it is the same as if the configuration is empty"